lib/carrier/messaging/tracker.ex

```diff
@@ -0,0 +1,179 @@
+  def add_reply_endpoint(%__MODULE__{reply_endpoints: reps}=tracker, subscriber) do
+    case Map.get(reps, subscriber) do
+      nil ->
+        topic = make_reply_endpoint_topic()
+        reps = reps
+        |> Map.put(subscriber, topic)
+        |> Map.put(topic, subscriber)
+        {maybe_monitor_subscriber(%{tracker | reply_endpoints: reps}, subscriber), topic}
+      topic ->
+        {tracker, topic}
+    end
+  end
+
+  @spec get_reply_endpoint(Tracker.t, pid()) :: String.t | nil
+  def get_reply_endpoint(%__MODULE__{reply_endpoints: reps}, subscriber) do
+    Map.get(reps, subscriber)
+  end
+
+  @spec add_subscription(Tracker.t, String.t, pid()) :: Tracker.t
+  def add_subscription(%__MODULE__{subscriptions: subs}=tracker, topic, subscriber) do
+    subs = case Map.get(subs, topic) do
+             nil ->
+               Map.put(subs, topic, {subscription_matcher(topic), [subscriber]})
+             {matcher, subscribed} ->
+               Map.put(subs, topic, {matcher, Enum.uniq([subscriber|subscribed])})
+           end
+    maybe_monitor_subscriber(%{tracker | subscriptions: subs}, subscriber)
+  end
+
+  @spec del_subscription(Tracker.t, String.t, pid()) :: {Tracker.t, boolean()}
+  def del_subscription(%__MODULE__{subscriptions: subs, unused_topics: ut}=tracker, topic, subscriber) do
+    case Map.get(subs, topic) do
+      {_matcher, [^subscriber]} ->
+        subs = Map.delete(subs, topic)
+        {maybe_unmonitor_subscriber(%{tracker | subscriptions: subs, unused_topics: Enum.uniq([topic|ut])}, subscriber), true}
+      {matcher, subscribed} ->
+        case List.delete(subscribed, subscriber) do
+          ^subscribed ->
+            {tracker, false}
+          updated ->
+            {maybe_unmonitor_subscriber(%{tracker | subscriptions: Map.put(subs, topic, {matcher, updated})}, subscriber), true}
+        end
+      nil ->
+        {tracker, false}
+    end
+  end
+
+  @spec find_reply_subscriber(Tracker.t, String.t) :: pid() | nil
+  def find_reply_subscriber(%__MODULE__{reply_endpoints: reps}, topic) do
+    Map.get(reps, topic)
+  end
+
+  @spec find_subscribers(Tracker.t, String.t) :: [] | [pid()]
+  def find_subscribers(%__MODULE__{subscriptions: subs}, topic) do
+    Enum.reduce(subs, [], &(find_matching_subscriptions(&1, topic, &2)))
+  end
+
+  @spec del_subscriber(Tracker.t, pid()) :: Tracker.t
+  def del_subscriber(tracker, subscriber) do
+    tracker
+    |> del_reply_endpoint(subscriber)
+    |> del_all_subscriptions(subscriber)
+    |> unmonitor_subscriber(subscriber)
+  end
+
+  @spec unused?(Tracker.t) :: boolean()
+  def unused?(%__MODULE__{reply_endpoints: reps, subscriptions: subs, monitors: monitors}) do
+    Enum.empty?(reps) and Enum.empty?(subs) and Enum.empty?(monitors)
+  end
+
+  @spec get_and_reset_unused_topics(Tracker.t) :: {Tracker.t, [String.t]}
+  def get_and_reset_unused_topics(tracker) do
+    {%{tracker | unused_topics: []}, tracker.unused_topics}
+  end
```

Comment:
Which would require external callers to directly access struct fields making it harder to change behavior in the future.